titus_isolate/event/oversubscribe_event_handler.py

```diff
@@ -0,0 +1,280 @@
+        self.__cpu_usage_predictor_manager = get_cpu_usage_predictor_manager()
+
+        self.__node_name = self.__config_manager.get_str(EC2_LOCAL_IPV4)
+        kubeconfig = get_kubeconfig_path()
+        self.__core_api = kubernetes.client.CoreV1Api(
+            kubernetes.config.new_client_from_config(config_file=kubeconfig))
+        self.__ext_api = kubernetes.client.ApiextensionsV1beta1Api(
+            kubernetes.config.new_client_from_config(config_file=kubeconfig))
+        self.__custom_api = kubernetes.client.CustomObjectsApi(
+            kubernetes.config.new_client_from_config(config_file=kubeconfig))
+
+    def set_registry(self, registry):
+        self.__reg = registry
+
+    def report_metrics(self, tags):
+        self.__reg.gauge(OVERSUBSCRIBE_FAIL_COUNT, tags).set(self.get_fail_count())
+        self.__reg.gauge(OVERSUBSCRIBE_SKIP_COUNT, tags).set(self.get_skip_count())
+        self.__reg.gauge(OVERSUBSCRIBE_SUCCESS_COUNT, tags).set(self.get_success_count())
+        self.__reg.gauge(OVERSUBSCRIBE_RECLAIMED_CPU_COUNT, tags).set(self.get_reclaimed_cpu_count())
+
+    def get_fail_count(self):
+        return self.__fail_count
+
+    def get_skip_count(self):
+        return self.__skip_count
+
+    def get_success_count(self):
+        return self.__success_count
+
+    def get_reclaimed_cpu_count(self):
+        return self.__reclaimed_cpu_count
+
+    def handle(self, event):
+        if not self.__relevant(event):
+            return
+
+        self.handling_event(event, 'oversubscribing workloads')
+
+        log.info('cleaning up old opportunistic resources')
+        clean_count = self.__cleanup()
+        log.info('cleaned up %d old opportunistic resources', clean_count)
+
+        cpu_usage = self.__workload_monitor_manager.get_cpu_usage(seconds=3600, agg_granularity_secs=60)
+        pred_env = PredEnvironment(self.__config_manager.get_region(), self.__config_manager.get_environment(),
+                                   datetime.utcnow().hour)
+
+        if self.__is_window_active():
+            self.__skip_count += 1
+            self.handled_event(event, 'skipping oversubscribe - a window is currently active')
+            return
+
+
+        workload_count = 0
+        free_cpu_count = 0
+        # we calculate the window before we send the request to ensure we're not going over our 10 minute mark
+        start = datetime.utcnow()
+        end = start + timedelta(minutes=10)
+        for workload in self.workload_manager.get_workloads():
+            # skip if an opportunistic workload already exists on this agent. this case only happens when an
+            # opportunistic workload outlives the window it was created for.
+            # TODO optimize by subtracting opportunistic CPUs used by opportunistic workloads instead of skipping.
+            # TODO optimize by reclaiming unused real CPUs from partially opportunistic workloads.
+            if workload.is_opportunistic():
```

Comment:
1. I would factor this out into a method which is something like:
```python
def is_oversubscribable(w: Workload) -> bool:
    pass
```
2. Perhaps we should not entirely abort oversubscription when a workload outlives it's intended lifetime, just skip it for oversubscription.  So `continue` below instead of `return`.

Reply:
addressed the TODOs to reclaim unused real CPUs as well as subtracting already oversubscribed cpus from the total. also added the `is_oversubscribable` method.